### PR TITLE
Upgrade protobuf

### PR DESCRIPTION
**Pull request by d61h6k4**
**Proposed changes**:
The module `tensorflow-hub` is currently not compatible with `protobuf-4.x` (meaning, authors of tensorflow-hub compiled their protobuf files with protoc compatible with protobuf-3.x and haven't rebuilt files with protoc compatible with protobuf-4.x). 

The Protobuf team suggests solving the problem 
1. by downgrading Protobuf (We are unable to downgrade to protobuf 3.23.3 as it leads to the failure of TensorFlow importation on the Docker image on the m1 chip.) 
2. by setting the environment variable `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python` 

This PR implements the second option.

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -153,38 +153,47 @@ else
 endif
 
 test-cli: PYTEST_MARKER=category_cli and (not flaky)
+test-cli: PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=upb
```
Comment (vcidst) [inline Makefile:156]: Your PR description mentions a different environment variable value `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python`, typo?

Comment (d61h6k4) [reply Makefile:156]: Not exactly; we are obliged to use slow Python implementation only when we work with TensorFlow-hub, which happens only in test-nlu-featurizers. We can use fast C implementation for all other cases, which is upb.

File: docs/docs/components.mdx [hunk 2]
```diff
@@ -493,7 +493,8 @@ Note: The `feature-dimension` for sequence and sentence features does not have t
   :::
 
   :::note
-  To use `ConveRTFeaturizer`, install Rasa with `pip3 install rasa[convert]`.
+  To use `ConveRTFeaturizer`, install Rasa with `pip3 install rasa[convert]` and set environment variable
+  `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python`.
 
   Note that this component cannot currently run on MacOS using M1 / M2 architecture.
```
[the remark this answers is not in the document]
Comment (d61h6k4) [reply docs/docs/components.mdx:498]: Good point; there is no extra section `convert` in `pyproject.toml`. I removed the pip install part.

File: rasa/utils/train_utils.py [hunk 3]
```diff
@@ -174,23 +172,35 @@ def update_evaluation_parameters(config: Dict[Text, Any]) -> Dict[Text, Any]:
 
 
 def load_tf_hub_model(model_url: Text) -> Any:
-    """Load model from cache if possible, otherwise from TFHub"""
+    """Load model from cache if possible, otherwise from TFHub."""
+    import os
+
+    # Once `tensorflow-hub` have rebuilt their `_pb2.py` files with protobuf-4.x
+    # this can be removed.
+    if os.getenv("PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION", "ubp") != "python":
```
Comment (radovanZRasa) [inline rasa/utils/train_utils.py:180]: PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION should be a constant in related constants.py.

File: tests/nlu/featurizers/test_convert_featurizer.py [hunk 4]
```diff
@@ -313,10 +305,31 @@ def test_raise_wrong_model_file(
...
 
     assert "neither a valid remote URL nor a local directory" in str(excinfo.value)
+
+
+@pytest.mark.skip_on_windows
+def test_raise_tensorflow_hub_incompatible(
+    create_or_load_convert_featurizer: Callable[[Dict[Text, Any]], ConveRTFeaturizer],
+    monkeypatch: MonkeyPatch,
+):
+    os.environ["PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION"] = "cpp"
```
Comment (radovanZRasa) [inline tests/nlu/featurizers/test_convert_featurizer.py:321]: We should use monkeypatch.setenv("PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION", "cpp").

**Review by radovanZRasa — APPROVED** [review]

File: rasa/utils/train_utils.py [hunk 5]
```diff
@@ -174,23 +173,35 @@ def update_evaluation_parameters(config: Dict[Text, Any]) -> Dict[Text, Any]:
 
 
 def load_tf_hub_model(model_url: Text) -> Any:
-    """Load model from cache if possible, otherwise from TFHub"""
+    """Load model from cache if possible, otherwise from TFHub."""
+    import os
+
+    # Once `tensorflow-hub` have rebuilt their `_pb2.py` files with protobuf-4.x
+    # this can be removed.
```
Comment (m-vdb) [inline rasa/utils/train_utils.py:180]: is there an issue / PR we can point to here, for reference?

File: docs/docs/components.mdx [hunk 6]
```diff
@@ -493,7 +493,7 @@ Note: The `feature-dimension` for sequence and sentence features does not have t
   :::
```
Comment (m-vdb) [inline docs/docs/components.mdx:493]: @d61h6k4 @twerkmeister do we have a place in the docs where we talk about M1 support more broadly? If we don't, I think we should add it (not sure where, maybe you have an idea?) and mention that full M1 support works from 3.6.2 onwards.

**Comment by d61h6k4** [comment]
@radovanZRasa @m-vdb By removing `tensorflow-io-gcs-filesystem` from constraints, I was able to simplify PR a lot, but we need to see that it doesn't break windows setup.

**Comment by d61h6k4** [comment]
@m-vdb all tests green, ready to merge

**Comment by d61h6k4** [comment]
@m-vdb I did one manual test.

Steps to reproduce:
* Build the docker image locally `PLATFORM=linux/arm64 make build-docker`
* Train concertbot
```
rasa/examples/concertbot on  ENG-284-upgrade-protobuf [$?] via 🐍  v3.10.12 (rasa-py3.10) on ☁  d.petrov@rasa.com took 47s 
❯ docker run -it --platform=linux/arm64 -v `pwd`:/app rasa:localdev train
```
Concertbot contains DIET and TED.

**Review by m-vdb — APPROVED** [review]
🚀 good to go on my end! thanks Danil 🙏🏻

**Pull request merged**